nflow-tests/src/main/java/io/nflow/tests/demo/workflow/BulkWorkflow.java

```diff
@@ -0,0 +1,137 @@
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import io.nflow.engine.service.WorkflowInstanceService;
+import io.nflow.engine.workflow.definition.NextAction;
+import io.nflow.engine.workflow.definition.StateExecution;
+import io.nflow.engine.workflow.definition.StateVar;
+import io.nflow.engine.workflow.definition.WorkflowDefinition;
+import io.nflow.engine.workflow.definition.WorkflowState;
+import io.nflow.engine.workflow.definition.WorkflowStateType;
+import io.nflow.engine.workflow.instance.WorkflowInstance;
+import io.nflow.engine.workflow.instance.WorkflowInstance.WorkflowInstanceStatus;
+
+/**
+ * Bulk child workflow executor that does not overflow the system.
+ */
+@Component
+public class BulkWorkflow extends WorkflowDefinition<BulkWorkflow.State> {
+
+  public static final String BULK_WORKFLOW_TYPE = "bulk";
+
+  private final static EnumSet<WorkflowInstanceStatus> RUNNING_STATES = complementOf(EnumSet.of(finished, created));
+
+  private static final Logger logger = getLogger(BulkWorkflow.class);
+
+  @Inject
+  WorkflowInstanceService instanceService;
+
+  public enum State implements WorkflowState {
+    splitWork(start), waitForChildrenToFinish(normal), done(end), error(manual);
+
+    private WorkflowStateType type;
+
+    State(WorkflowStateType type) {
+      this.type = type;
+    }
+
+    @Override
+    public WorkflowStateType getType() {
+      return type;
+    }
+
+    @Override
+    public String getDescription() {
+      return name();
+    }
+  }
+
+  protected BulkWorkflow(String type) {
+    super(type, splitWork, error);
+    setDescription("Executes child workflows in bulk but gracefully without effecting non-bulk tasks.");
+    permit(splitWork, waitForChildrenToFinish);
+    permit(waitForChildrenToFinish, done);
+  }
+
+  public BulkWorkflow() {
+    this(BULK_WORKFLOW_TYPE);
+  }
+
+  public NextAction splitWork(StateExecution execution, @StateVar(value = "requestData", readOnly = true) JsonNode data) {
+    boolean childrenFound = splitWorkImpl(execution, data);
+    if (childrenFound) {
+      return moveToState(waitForChildrenToFinish, "Running");
+    }
+    return retryAfter(now().plusMinutes(1), "Waiting for child workflows");
+  }
+
+  protected boolean splitWorkImpl(StateExecution execution, @SuppressWarnings("unused") JsonNode data) {
+    if (execution.getAllChildWorkflows().isEmpty()) {
+      throw new RuntimeException(
+          "No child workflows found - either add them before starting the parent or implement splitWorkflowImpl");
+    }
+    return true;
+  }
+
+  public NextAction waitForChildrenToFinish(StateExecution execution,
+      @StateVar(value = "concurrency", readOnly = true) int concurrency) {
+    List<WorkflowInstance> childWorkflows = execution.getAllChildWorkflows();
+    long running = childWorkflows.stream().filter(this::isRunning).count();
+    long completed = childWorkflows.stream().filter(this::isFinished).count();
```

Comment:
This will now loop through the upto 100k size children twice. We could do it with streams by categorizing on a Map<Type,Count> with groupBy operation. We just need a new enum for that I think.

Reply:
Just reverted to how it was